### ISSUE TITLE
always_specify_types - except for final and const

### ISSUE THREAD

**Issue by zoechi** [dart-lang/linter#155]
I really like this rule, but AFAIK the analyzer uses the inferred type anyway (not 100% sure, can you confirm?) and there is no way to assign a value violating this type later.
Either the Flutter team can be convinced that type annotations are redundant here, or please create a deviation of this rule where `const` and `final` field declarations are excluded.
I had the impression not using type annotations for `final` and `const` was according to the Dart style guide, but couldn't find it mentioned.

**Comment by bwilkerson** [comment]
The analyzer uses the static type for purposes of generating errors and warnings. Strong mode changes the static type from what is defined by the specification, so if you're using strong mode then it would use the inferred type. But if you're not using strong mode then the inferred type is only used for things like code completion (and there's currently a bug that analyzer isn't actually inferring types for these variables).

It is possible that the Flutter team will choose to relax this requirement after they have more experience with strong mode, but avoiding errors is not the only reason for specifying types (and in my opinion not even the strongest). I use types to express my intent so that the tools can verify that I have not violated my own intentions. I don't get that level of assurance if I let the tools guess at my intent. So I think it makes sense to have a version of this rule that is strict.

But yes, for those who prefer to not provide type annotations for final and const variables, it makes perfect sense to have a variation of this rule.

Now we just have to decide how to do that. Do we have two rules: always_specify_types and always_specify_nonfinal_types? Or do we allow rules to be configured:

    always_specify_types:
      including_final: false

**Comment by zoechi** [comment]
@bwilkerson I fully agree with your explanation about intent.

I just find 
```dart
final MyClass<String,SomeOtherClass> someVar = new MyClass<String,SomeOtherClass>();
```
needlessy verbose.
When I usually omit the type after `final` in such obvious cases like 
```dart
final someVar = new MyClass<String,SomeOtherClass>();
// or
final someVar = 'abc';
```
it would make deviations much better visible
```dart
final MyInterface<String,SomeOtherClass> someVar = new MyClass<String,SomeOtherClass>();
```

---
I'm in favor of different rules, they can of course share code if this makes sense from this perspective.
I'd prefer to have rulesets (or tags on rules), like `dart style guide`, `flutter`, ... and a way to enable a rulesets and selectively disable single rules from such a set and also enable single rules independent of the set like

```yaml
rules:
  - dart_style_guide:
    -exclude:
      - unnecessary_brace_in_string_interp 
  - avoid_as
```
while one rule can be in more than one set (or have more than one tag).
See also #103

**Comment by kasperpeulen** [comment]
> I had the impression not using type annotations for final and const was according to the Dart style guide, but couldn't find it mentioned.

I think the dart style guide actually says to not use type annotations for any local variable (regardless of `final`, `const`. The dart style guide basically says that

```
MyClass<String,SomeOtherClass> someVar = new MyClass<String,SomeOtherClass>();
```

is also needlessly verbose.

**Comment by pq** [comment]
@zoechi: in the interest of managing our issue backlog I'm doing some cleanup and closing this one.  Feel free to re-open or create a fresh one that reflects any possible nuances w/ Dart2.

Thanks!
